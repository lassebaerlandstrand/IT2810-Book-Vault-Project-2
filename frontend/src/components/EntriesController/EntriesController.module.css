.select {
  max-width: 5rem;

  label {
    font-size: 0.75rem;
    font-weight: 400;
  }
}
