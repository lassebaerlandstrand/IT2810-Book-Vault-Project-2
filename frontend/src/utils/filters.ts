import { SortBy, SortOrder } from '@/generated/graphql';

export const DEFAULT_FILTERS = {
  sortBy: SortBy.BookName,
  sortOrder: SortOrder.Asc,
  selectedGenres: [],
  selectedPublishers: [],
  selectedAuthors: [],
  selectedBeforeDate: new Date(2020, 11, 31),
  selectedAfterDate: new Date(1900, 0, 1),
  selectedMinPages: undefined,
  selectedMaxPages: undefined,
  selectedMinRating: undefined,
};

export const getFilterParams = (searchParams: URLSearchParams) => {
  const sortBy = (searchParams.get('sortBy') as SortBy) ?? DEFAULT_FILTERS.sortBy;
  const sortOrder = (searchParams.get('sortOrder') as SortOrder) ?? DEFAULT_FILTERS.sortOrder;
  const authors = searchParams.getAll('authors') ?? DEFAULT_FILTERS.selectedAuthors;
  const publishers = searchParams.getAll('publishers') ?? DEFAULT_FILTERS.selectedPublishers;
  const genres = searchParams.getAll('genres') ?? DEFAULT_FILTERS.selectedGenres;
  const beforeDate = searchParams.get('beforeDate')
    ? new Date(searchParams.get('beforeDate') ?? '')
    : DEFAULT_FILTERS.selectedBeforeDate;
  const afterDate = searchParams.get('afterDate')
    ? new Date(searchParams.get('afterDate') ?? '')
    : DEFAULT_FILTERS.selectedAfterDate;
  const minPages =
    parseInt(searchParams.get('minPages') ?? '', 10) || DEFAULT_FILTERS.selectedMinPages;
  const maxPages =
    parseInt(searchParams.get('maxPages') ?? '', 10) || DEFAULT_FILTERS.selectedMaxPages;
  const minRating =
    parseInt(searchParams.get('minRating') ?? '', 10) || DEFAULT_FILTERS.selectedMinRating;

  return {
    sortBy,
    sortOrder,
    authors,
    publishers,
    genres,
    beforeDate,
    afterDate,
    minPages,
    maxPages,
    minRating,
  };
};
