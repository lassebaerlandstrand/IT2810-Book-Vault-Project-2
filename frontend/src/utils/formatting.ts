/**
 * Formats a number by adding spaces as thousand separators.
 *
 * @param {number} number - The input number to format
 * @returns {string} The formatted number string with space separators
 *
 * @example
 * // returns '1 000'
 * formatNumberWithSpaces(1000);
 *
 * @example
 * // returns '1 000 000'
 * formatNumberWithSpaces(1000000);
 */
export const formatNumberWithSpaces = (number: number) =>
  number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ' ');

// The regular expression used in the replace method:
// \B - matches a non-word boundary to avoid adding spaces at the start of the string.
// (?=(\d{3})+(?!\d)) - positive lookahead to find positions followed by a group of three digits not followed by another digit.
