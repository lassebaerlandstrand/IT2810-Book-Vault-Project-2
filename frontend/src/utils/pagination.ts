const DEFAULT_PAGE = 1;
const DEFAULT_LIMIT = 25;
const LIMIT_OPTIONS = ['10', '25', '50', '100'];

export const getPaginationParams = (searchParams: URLSearchParams) => {
  const page = Number(searchParams.get('page')) || DEFAULT_PAGE;
  const limit = Number(searchParams.get('limit')) || DEFAULT_LIMIT;

  return { page, limit, DEFAULT_PAGE, DEFAULT_LIMIT, LIMIT_OPTIONS };
};
