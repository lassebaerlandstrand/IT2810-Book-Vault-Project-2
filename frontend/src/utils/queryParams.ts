import { SetURLSearchParams } from 'react-router-dom';

/**
 * Updates the URL search parameters based on the provided key and value.
 * Handles different types of values (single string, array of strings, or undefined).
 * This function handles consecutive calls.
 *
 * @param {SetURLSearchParams} setSearchParams - Function to update the URL search parameters
 * @param {string} key - The key of the query parameter to update
 * @param {string | string[] | undefined} value - The value to set for the query parameter
 *
 * @example
 * updateQueryParams(setSearchParams, 'search', 'newSearchQuery')
 * updateQueryParams(setSearchParams, 'genres', ['tag1', 'tag2'])
 * updateQueryParams(setSearchParams, 'page', '5')
 * updateQueryParams(setSearchParams, 'page', undefined) // Removes the key
 *
 */
export const updateQueryParams = (
  setSearchParams: SetURLSearchParams,
  key: string,
  value: string | string[] | undefined
) => {
  setSearchParams((params) => {
    if (Array.isArray(value)) {
      params.delete(key);
      value.forEach((v) => params.append(key, v));
    } else if (value === undefined || value === '') {
      params.delete(key);
    } else {
      params.set(key, value);
    }
    return params;
  });
};
