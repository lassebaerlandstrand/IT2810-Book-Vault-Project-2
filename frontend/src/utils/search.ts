/**
 * Extracts the search parameter from the URL.
 * This function ensures that the search parameters are correctly parsed and returned in a consistent format.
 *
 * @param {URLSearchParams} searchParams - URL search parameters containing search criteria
 * @returns {Object} Search parameters
 *
 */
export const getSearchParams = (searchParams: URLSearchParams) => {
  const searchValue = searchParams.get('search') || '';
  return { searchValue };
};
