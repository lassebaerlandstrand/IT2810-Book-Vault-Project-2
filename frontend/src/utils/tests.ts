/**
 * Removes random attributes generated by Mantine from DOM elements.
 * This is necessary to ensure consistent snapshots in tests, as Mantine uses random IDs.
 */
export const removeMantineRandomAttributes = () => {
  const attributesToRemove = [
    ...document.body.querySelectorAll('div [id^="mantine"]'),
    ...document.body.querySelectorAll('div [for^="mantine"]'),
  ];
  attributesToRemove.forEach((element) => {
    element.removeAttribute('for');
    element.removeAttribute('id');
    element.removeAttribute('class');
    element.removeAttribute('aria-describedby');
  });
};
