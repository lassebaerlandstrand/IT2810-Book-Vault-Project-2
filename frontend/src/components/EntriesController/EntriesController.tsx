import { useSearchParams } from 'react-router-dom';
import { Select } from '@mantine/core';
import { getPaginationParams } from '@/utils/pagination';
import styles from './EntriesController.module.css';

const EntriesController = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const { limit, DEFAULT_LIMIT, LIMIT_OPTIONS } = getPaginationParams(searchParams);

  const handleEntiresLimitChange = (newLimit: string | null) => {
    const newParams = new URLSearchParams(searchParams.toString());
    newParams.set('limit', newLimit ?? DEFAULT_LIMIT.toString());
    setSearchParams(newParams);
  };

  return (
    <Select
      label="Entries"
      data={LIMIT_OPTIONS}
      allowDeselect={false}
      value={limit.toString()}
      onChange={handleEntiresLimitChange}
      aria-label="Number of entries"
      className={styles.select}
      data-testid="entries-select"
    />
  );
};

export default EntriesController;
