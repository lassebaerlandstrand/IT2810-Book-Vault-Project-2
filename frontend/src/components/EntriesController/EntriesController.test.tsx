import { fireEvent, render, screen } from '@test-utils';
import { MemoryRouter } from 'react-router-dom';
import * as reactRouterDom from 'react-router-dom'; // This was the only was to get spyOn to work with useSearchParams

import { removeMantineRandomAttributes } from '@/utils/tests';
import EntriesController from './EntriesController';

vi.mock('@/utils/pagination', () => ({
  getPaginationParams: vi.fn(() => ({
    page: 1,
    limit: 10,
    DEFAULT_LIMIT: 25,
    LIMIT_OPTIONS: ['10', '25', '50', '100'],
  })),
}));

describe('EntriesController', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders the select component with correct initial value', () => {
    render(
      <MemoryRouter>
        <EntriesController />
      </MemoryRouter>
    );

    expect(screen.getByText('Entries')).toBeInTheDocument();
    expect(screen.getByTestId('entries-select')).toHaveValue('10');
  });

  test('updates the search params when a new limit is selected', () => {
    const mockSetSearchParams = vi.fn();
    vi.spyOn(reactRouterDom, 'useSearchParams').mockReturnValue([
      new URLSearchParams(),
      mockSetSearchParams,
    ]);

    render(
      <MemoryRouter>
        <EntriesController />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByText('50'));

    expect(mockSetSearchParams).toHaveBeenCalledWith(expect.any(URLSearchParams));
    const params = mockSetSearchParams.mock.calls[0][0];
    expect(params.get('limit')).toBe('50');
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<EntriesController />);
    removeMantineRandomAttributes();
    expect(asFragment()).toMatchSnapshot();
  });
});
