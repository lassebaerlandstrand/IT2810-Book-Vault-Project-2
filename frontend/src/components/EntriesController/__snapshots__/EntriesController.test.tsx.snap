// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EntriesController > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;--mantine-color-logo-theme-0: #fefefe;--mantine-color-logo-theme-1: #cbcfd2;--mantine-color-logo-theme-2: #97a0a6;--mantine-color-logo-theme-3: #7d8991;--mantine-color-logo-theme-4: #63717b;--mantine-color-logo-theme-5: #495a65;--mantine-color-logo-theme-6: #2f434f;--mantine-color-logo-theme-7: #132937;--mantine-color-logo-theme-8: #0b202e;--mantine-color-logo-theme-9: #061017;}:root[data-mantine-color-scheme="dark"]{--mantine-color-logo-theme-text: var(--mantine-color-logo-theme-4);--mantine-color-logo-theme-filled: var(--mantine-color-logo-theme-8);--mantine-color-logo-theme-filled-hover: var(--mantine-color-logo-theme-9);--mantine-color-logo-theme-light: rgba(47, 67, 79, 0.15);--mantine-color-logo-theme-light-hover: rgba(47, 67, 79, 0.2);--mantine-color-logo-theme-light-color: var(--mantine-color-logo-theme-3);--mantine-color-logo-theme-outline: var(--mantine-color-logo-theme-4);--mantine-color-logo-theme-outline-hover: rgba(99, 113, 123, 0.05);}:root[data-mantine-color-scheme="light"]{--mantine-color-logo-theme-text: var(--mantine-color-logo-theme-filled);--mantine-color-logo-theme-filled: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-filled-hover: var(--mantine-color-logo-theme-7);--mantine-color-logo-theme-light: rgba(47, 67, 79, 0.1);--mantine-color-logo-theme-light-hover: rgba(47, 67, 79, 0.12);--mantine-color-logo-theme-light-color: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-outline: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-outline-hover: rgba(47, 67, 79, 0.05);}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <div
    class="_select_5bf5c3 m_46b77525 mantine-InputWrapper-root mantine-Select-root"
  >
    <label
      class="m_8fdc1311 mantine-InputWrapper-label mantine-Select-label"
    >
      Entries
    </label>
    <div
      class="m_6c018570 mantine-Input-wrapper mantine-Select-wrapper"
      data-pointer="true"
      data-variant="default"
      data-with-right-section="true"
      style="--input-right-section-pointer-events: none;"
    >
      <input
        aria-haspopup="listbox"
        aria-invalid="false"
        aria-label="Number of entries"
        autocomplete="off"
        class="m_8fb7ebe7 mantine-Input-input mantine-Select-input"
        data-testid="entries-select"
        data-variant="default"
        readonly=""
        value="10"
      />
      <div
        class="m_82577fc2 mantine-Input-section mantine-Select-section"
        data-position="right"
      >
        <svg
          class="m_2943220b mantine-ComboboxChevron-chevron"
          data-combobox-chevron="true"
          fill="none"
          viewBox="0 0 15 15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
  </div>
  <input
    type="hidden"
    value="10"
  />
</DocumentFragment>
`;
